idea/idea-frontend-fir/src/org/jetbrains/kotlin/idea/references/FirReferenceResolveHelper.kt

```diff
@@ -374,42 +375,104 @@ internal object FirReferenceResolveHelper {
         fir: FirResolvedQualifier,
         expression: KtSimpleNameExpression,
         session: FirSession,
-        symbolBuilder: KtSymbolByFirBuilder,
-        analysisSession: KtFirAnalysisSession
+        symbolBuilder: KtSymbolByFirBuilder
     ): Collection<KtSymbol> {
-        // TODO refactor that block
-        val classId = fir.classId ?: return listOfNotNull(getPackageSymbolFor(expression, symbolBuilder, forQualifiedType = false))
+        val referencedSymbol = if (fir.resolveToCompanionObject) {
+            (fir.symbol?.fir as? FirRegularClass)?.companionObject?.symbol
+        } else {
+            fir.symbol
+        }
+        if (referencedSymbol == null) {
+            // If referencedSymbol is null, it means the reference goes to a package.
+            val parent = expression.parent as? KtDotQualifiedExpression ?: return emptyList()
+            val fqNameSegments =
+                when (expression) {
+                    parent.selectorExpression -> parent.fqNameSegments() ?: return emptyList()
+                    parent.receiverExpression -> listOf(expression.getReferencedName())
+                    else -> return emptyList()
+                }
+            return listOfNotNull(symbolBuilder.createPackageSymbolIfOneExists(FqName.fromSegments(fqNameSegments)))
+        }
+        val referencedClass = referencedSymbol.fir
+        val referencedSymbolsByFir = listOfNotNull(symbolBuilder.buildSymbol(referencedClass))
+        if (fir.source.psi === expression) return referencedSymbolsByFir
+
+        if (referencedClass.isLocal) {
+            // TODO: handle local classes after KT-47135 is fixed
+            return referencedSymbolsByFir
+        } else {
+            // In the code below, we always maintain the contract that `classId` and `qualifiedAccess` should stay "in-sync", i.e. they
+            // refer to the same class.
+            var qualifiedAccess = fir.source.psi?.let {
+                // In some cases, the source from FIR is a KtDotQualifiedExpression and in some other cases, it's the selectorExpression of
+                // a KtDotQualifiedExpression
+                it as? KtDotQualifiedExpression ?: it.parent as? KtDotQualifiedExpression
```

Comment:
A little confused by this now after your reply to the above comment. My understanding from your reply is at this point that the source of `fir` should be a `KtDotQualifiedExpression`. Please correct if I'm wrong, and if so, could you put some example cases in the comment here.

Reply:
I think your understanding is correct. I can add an assert and remove the defensive `it.parent` check here. If later it turns out this assumption is false. We can add it back.